### ISSUE TITLE
False positive `unused_variables` when the variable is used in conditional code

### ISSUE THREAD

**Issue by daira** [rust-lang/rust-clippy#15906]
### Summary

Consider code like this (simplified from real code [here](https://github.com/zcash/librustzcash/pull/1999/files#diff-57d91750cf8573e5cfebdee8bd31b2d635cb02adf7f5e5fdd20ff2253e43782fR1037-R1055)):

```rust
if let Some(v) = ... {
    // side effects

    #[cfg(feature = "some-feature")]
    code_that_uses(v);
}
```

It is not unexpected that this causes an `unused_variables` warning when "some-feature" is disabled, but I think it shouldn't. In fact this is quite a common pattern and is almost never a real bug, in my experience. Note that, even if `v` isn't used in a particular configuration, we must name it something because it is used in *some* configurations, and in any case the `if let Some` can't be omitted.

There are two ways to suppress the warning.
* We can call the variable by a name starting `_`. But I find that ugly: I prefer to reserve `_` for unused parameters, because that genuinely is a little unusual and deserves to be called out, whereas having code that conditionally uses a variable is (at least for the way we use features in my workplace) common.
* We can add `#[allow(unused_variables)]`. But that is verbose and arguably shouldn't be needed. The variable *is* used. It is not likely to be an error that it isn't *always* used, any more than it would be if we had written `if (some_feature) { code_that_uses(v); }`.

### Lint Name

`unused_variables`

### Reproducer

See above.

### Version

I saw it in `clippy 0.1.81` but I'm pretty sure it still does this in more recent versions.
